cxml/xml.go

```diff
@@ -0,0 +1,222 @@
+// into a series of outgoing tokens when traversing XML tree
+type TransformFunc func(parents *NodeList, in xml.Token) []xml.Token
+
+// ConditionFunc is accepted by some transform function generators
+// so they could be applied conditionaly
+type ConditionFunc func(parents *NodeList, in xml.Token) bool
+
+// TransformXML parses the XML tree, traverses it and calls TransformFunc
+// on each XML token, writing the output to the writer, resulting in a
+// transformed XML tree
+func TransformXML(r io.Reader, w io.Writer, fn TransformFunc, indent bool) error {
+	parser := xml.NewDecoder(r)
+	encoder := xml.NewEncoder(w)
+	if indent {
+		encoder.Indent("  ", "    ")
+	}
+	parentNodes := &NodeList{}
+	for {
+		token, err := parser.Token()
+		if err != nil {
+			if err != io.EOF {
+				return trace.Wrap(err)
+			}
+			break
+		}
+		for _, t := range fn(parentNodes, token) {
+			if err := encoder.EncodeToken(t); err != nil {
+				return err
+			}
+		}
+		switch e := token.(type) {
+		case xml.StartElement:
+			parentNodes.Push(e)
+		case xml.EndElement:
+			parentNodes.Pop()
+		}
+	}
+	encoder.Flush()
+	return nil
+}
+
+// SetAttribute is XML helper that allows to set attribute on a node
+func SetAttribute(e xml.StartElement, name, value string) xml.StartElement {
+	if len(e.Attr) != 0 {
+		for i := range e.Attr {
+			if e.Attr[i].Name.Local == name {
+				e.Attr[i].Value = value
+				return e
+			}
+		}
+	} else {
+		e.Attr = append(e.Attr, xml.Attr{Name: xml.Name{Local: name}, Value: value})
+	}
+	return e
+}
+
+// ReplaceCDATAIf replaces CDATA value of the matched node
+// if the parent node name matches the name
+func ReplaceCDATAIf(val []byte, cond ConditionFunc) TransformFunc {
+	return func(parents *NodeList, in xml.Token) []xml.Token {
+		switch in.(type) {
+		case xml.CharData:
+			if cond(parents, in) {
+				data := make([]byte, len(val))
+				copy(data, val)
+				return []xml.Token{xml.CharData(data)}
+			}
+		}
+		return []xml.Token{in}
+	}
+}
+
+// InjectNodes injects nodes at the end of the tag that matches name
+func InjectNodesIf(nodes []xml.Token, cond ConditionFunc) TransformFunc {
+	return func(parents *NodeList, in xml.Token) []xml.Token {
+		if cond(parents, in) {
+			return append(nodes, in)
+		}
+		return []xml.Token{in}
+	}
+}
+
+// ReplaceAttribute replaces the attribute of the first node that matches the name
+func ReplaceAttributeIf(attrName, attrValue string, cond ConditionFunc) TransformFunc {
+	return func(parents *NodeList, in xml.Token) []xml.Token {
+		switch t := in.(type) {
+		case xml.StartElement:
+			if cond(parents, in) {
+				e := xml.StartElement(t)
+				return []xml.Token{SetAttribute(e, attrName, attrValue)}
+			}
+		}
+		return []xml.Token{in}
+	}
+}
+
+// TrimSpace is a transformer function that replaces CDATA with blank
+// characters with empty strings
+func TrimSpace(parents *NodeList, in xml.Token) []xml.Token {
+	switch t := in.(type) {
+	case xml.CharData:
+		if strings.TrimSpace(string(t)) == "" {
```

Comment:
There is a https://golang.org/pkg/bytes/#TrimSpace - this will avoid a redundant string conversion here.